hicommonwealth/commonwealth
# Discord Bot: Listener + Consumer + UI changes

## Link to Issue
Closes: #4266, Closes #4267, Closes #4268, Closes #4374, Closes #4469

## Description of Changes
#### discord-bot package
- Discord bot Listener
- Discord bot Consumer
#### common-common package
- Update Rascal config to include DiscordMessage Queue on SnapshotListener exchanges(per conversation with @timolegros we decided we should not create new exchanges for every queue we add)
#### Commonwealth package
- CW database Migrations
   - Creates discord_meta columns on Threads and Comments, channel_id on Topics
   - Creates Discord Bot user/profile
- Addition of Discord forum <-> Common Topic matching component on admin panel
    - Creates backend routes + updates Topic models to add channel_id to topic 
- Fixes/upgrades to discord admin functionality 
- Updates to `setDiscordConfig` Route to create a Address for each chain/community bot is added to
- New `/bot` prefixed routes for createThread/Comment 
   - Includes middleware which checks for `CW_BOT_KEY` and populates request with bot user if passing. Replaces JWT auth for bot routes
- Updates models and save logic to include discord_meta optionally 

## Test Plan
- See README in discord-bot package, set up listener and consumer with required env variables
- Start rabbitmq
- Start CW app(ensure `DISCORD_CLIENT_ID` `DISCORD_BOT_TOKEN` and `CW_BOT_KEY` are set to respective values(again see README)) 
- In a community where you're an admin, go to manage community and connect the bot to a test discord server(that includes a forum)
- Use the topic component in the admin panel to connect a forum to a topic
- Create a forum post in discord
- confirm post is created in community
- comment on post
- confirm comment is included in community 

## Deployment Plan

See Launch plan here: https://www.notion.so/buildcommon/Discobot-System-Design-Tech-Spec-533b1afd75f64feba038d3626d264ef4?pvs=4

## Other Considerations

 - discord_meta fields currently include: discord user data, channel_id, and message_id. These have been selected in order to support future expansion of the listener/consumers/apps features(these could be follow up tickets) such as 
     - Claiming discord users and converting threads author from bot to actual author
     - Reaction tracking(not very straightforward on the discord API side)
     - Update/delete functionality
- Before merging it may be worth confirming with product what the Discord Bot profile should look like. Currently has
    - Username: Discord Bot 
    - Address: 0xdiscordbot
    - permissions: admin(there is no actual user who can access this account so this should be safe) 
    - migrations/direct DB updates to profile can be used in the future to modify profile
